src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -314,280 +353,601 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+            {
+                localDbIdsList.Add((LazyStringValue)dbIds[i]);
             }
+
+            return localDbIdsList;
         }
 
-        public IEnumerable<string> GetCountersForDocument(DocumentsOperationContext context, string docId)
+        private bool InternalPutCounter(DocumentsOperationContext context, BlittableJsonReaderObject counters, string counterName,
+            DbIdsHolder localDbIds, BlittableJsonReaderArray sourceDbIds,
+            BlittableJsonReaderObject.RawBlob existingCounter, BlittableJsonReaderObject.RawBlob source)
         {
-            var table = new Table(CountersSchema, context.Transaction.InnerTransaction);
+            var existingCount = existingCounter.Length / SizeOfCounterValues;
+            var sourceCount = source.Length / SizeOfCounterValues;
+            var modified = false;
 
-            using (GetCounterPartialKey(context, docId, out var key))
+            for (var index = 0; index < sourceCount; index++)
             {
-                LazyStringValue prev = null;
-                foreach (var result in table.SeekByPrimaryKeyPrefix(key, Slices.Empty, 0))
-                {
-                    var current = ExtractCounterName(context, result.Value.Reader);
+                var sourceDbId = (LazyStringValue)sourceDbIds[index];
+                var sourceValue = &((CounterValues*)source.Ptr)[index];
 
-                    if (prev?.Equals(current) == true)
-                    {
-                        // already seen this one, skip it 
+                int localDbIdIndex = localDbIds.GetOrAddDbIdIndex(sourceDbId);
+
+                if (localDbIdIndex < existingCount)
+                {
+                    var localValuePtr = (CounterValues*)existingCounter.Ptr + localDbIdIndex;
+                    if (localValuePtr->Etag >= sourceValue->Etag)
                         continue;
-                    }
 
-                    yield return current;
+                    localValuePtr->Value = sourceValue->Value;
+                    localValuePtr->Etag = sourceValue->Etag;
 
-                    prev?.Dispose();
-                    prev = current;
+                    continue;
                 }
+
+                // counter doesn't have this dbId
+                modified = true;
+                AddPartialValueToExistingCounter(context, existingCounter, localDbIdIndex, sourceValue->Value, sourceValue->Etag);
+
+                existingCount = existingCounter.Length / SizeOfCounterValues;
             }
+
+            if (modified)
+            {
+                counters.Modifications = counters.Modifications ?? new DynamicJsonValue(counters);
+                counters.Modifications[counterName] = existingCounter;
+            }
+
+            return modified;
         }
 
-        public long? GetCounterValue(DocumentsOperationContext context, string docId, string counterName)
+        private void AddPartialValueToExistingCounter(DocumentsOperationContext context,
```

Comment:
This mutate the arg in a non obvious way, better to return `RawBlob` here

Reply:
done